File: src/hasoftware-library/src/hasoftware/api/MessageFactory.java
package hasoftware.api;

import hasoftware.api.messages.CurrentEventRequest;
import hasoftware.api.messages.CurrentEventResponse;
import hasoftware.api.messages.ErrorResponse;
import hasoftware.api.messages.InputMessageRequest;
import hasoftware.api.messages.InputMessageResponse;
import hasoftware.api.messages.LocationRequest;
import hasoftware.api.messages.LocationResponse;
import hasoftware.api.messages.LoginRequest;
import hasoftware.api.messages.LoginResponse;
import hasoftware.api.messages.NotifyRequest;
import hasoftware.api.messages.NotifyResponse;
import hasoftware.api.messages.OutputDeviceRequest;
import hasoftware.api.messages.OutputDeviceResponse;
import hasoftware.api.messages.OutputMessageRequest;
import hasoftware.api.messages.OutputMessageResponse;
import hasoftware.api.messages.PointRequest;
import hasoftware.api.messages.PointResponse;
import hasoftware.cdef.CDEFMessage;
import hasoftware.cdef.CDEFSystemFlags;

public class MessageFactory {

    public static Message decode(CDEFMessage cdefMessage) {
        Message message = null;
        int functionCode = cdefMessage.getU16(0);
        int systemFlags = cdefMessage.getU8(6);
        if (systemFlags == (CDEFSystemFlags.Error | CDEFSystemFlags.Response)) {
            message = new ErrorResponse(cdefMessage);
        } else {
            boolean isRequest = ((systemFlags & CDEFSystemFlags.Response) == 0);
            switch (functionCode) {
                case FunctionCode.Notify:
                    return (isRequest) ? new NotifyRequest(cdefMessage) : new NotifyResponse(cdefMessage);
                case FunctionCode.InputMessage:
                    return (isRequest) ? new InputMessageRequest(cdefMessage) : new InputMessageResponse(cdefMessage);
                case FunctionCode.OutputMessage:
                    return (isRequest) ? new OutputMessageRequest(cdefMessage) : new OutputMessageResponse(cdefMessage);
                case FunctionCode.OutputDevice:
                    return (isRequest) ? new OutputDeviceRequest(cdefMessage) : new OutputDeviceResponse(cdefMessage);
                case FunctionCode.Login:
                    return (isRequest) ? new LoginRequest(cdefMessage) : new LoginResponse(cdefMessage);
                case FunctionCode.Location:
                    return (isRequest) ? new LocationRequest(cdefMessage) : new LocationResponse(cdefMessage);
                case FunctionCode.Point:
                    return (isRequest) ? new PointRequest(cdefMessage) : new PointResponse(cdefMessage);
                case FunctionCode.CurrentEvent:
                    return (isRequest) ? new CurrentEventRequest(cdefMessage) : new CurrentEventResponse(cdefMessage);
            }
        }
        return message;
    }
}

File: src.net/hasoftware-library/Api/MessageFactory.cs
using hasoftware.Cdef;
using hasoftware.Messages;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace hasoftware.Api
{
    public class MessageFactory
    {
        public static Message Decode(CdefMessage cdefMessage)
        {
            Message message = null;
            int functionCode = cdefMessage.GetU16(0);
            int systemFlags = cdefMessage.GetU8(6);
            if (systemFlags == (CdefSystemFlags.Error | CdefSystemFlags.Response))
            {
                message = new ErrorResponse(cdefMessage);
            }
            else
            {
                bool isRequest = ((systemFlags & CdefSystemFlags.Response) == 0);
                switch (functionCode)
                {
                    case FunctionCode.Notify:
                        if (isRequest) message = new NotifyRequest(cdefMessage);
                        else message = new NotifyResponse(cdefMessage);
                        break;

                    case FunctionCode.InputMessage:
                        if (isRequest) message = new InputMessageRequest(cdefMessage);
                        else message = new InputMessageResponse(cdefMessage);
                        break;

                    case FunctionCode.OutputMessage:
                        if (isRequest) message = new OutputMessageRequest(cdefMessage);
                        else message = new OutputMessageResponse(cdefMessage);
                        break;

                    case FunctionCode.OutputDevice:
                        if (isRequest) message = new OutputDeviceRequest(cdefMessage);
                        else message = new OutputDeviceResponse(cdefMessage);
                        break;

                    case FunctionCode.Login:
                        if (isRequest) message = new LoginRequest(cdefMessage);
                        else message = new LoginResponse(cdefMessage);
                        break;

                    case FunctionCode.Location:
                        if (isRequest) message = new LocationRequest(cdefMessage);
                        else message = new LocationResponse(cdefMessage);
                        break;

                    case FunctionCode.Point:
                        if (isRequest) message = new PointRequest(cdefMessage);
                        else message = new PointResponse(cdefMessage);
                        break;

                    case FunctionCode.CurrentEvent:
                        if (isRequest) message = new CurrentEventRequest(cdefMessage);
                        else message = new CurrentEventResponse(cdefMessage);
                        break;
                }
            }
            return message;
        }
    }
}
